build.gradle

```diff
@@ -145,6 +142,8 @@ configure(javaProjects) {
   // ----------
 
   task sourcesJar(type: Jar, dependsOn: classes) {
+    duplicatesStrategy = DuplicatesStrategy.EXCLUDE
```

Comment:
Oh, interesting. Were we having duplicate entries? Which files? Or, was it that you just noticed a deprecation warning and found a way to resolve it?

Reply:
I get it on the `gax/src/main/resources/dependencies.properties` and only when running `./gradlew build publishToMavenLocal`. It looks like the file created by `build` overlaps with `sourcesJar`. If there is a better way to handle it, please let me know.

Reply:
Hmm... I think this is because now `gax/src/main/resources/dependencies.properties` is checked in as source. So I guess what's happening is that, because a final version is generated at `gax/build/resources/main/dependencies.properties`, now you have two `dependencies.properties` to include.

I think `DuplicatesStrategy.EXCLUDE` could be a problem, because it doesn't deterministically guarantee which file will win in the end. However, for a source JAR, I actually start to think there may be no reason to add this properties file? Not sure if it's the standard practice to add this kind of resource files into a source JAR. OTOH, since the file is part of the source set (it's under `src/main`), it also makes sense to include it, perhaps the one before processing with templating.

Let me try to find a way to include the original source file.

Reply:
Actually, it is correct that the error is due to the new `ax/src/main/resources/dependencies.properties`, but this is not because we have a processed file with the same filename under `gax/build/`.

```gradle
  task sourcesJar(type: Jar, dependsOn: classes) {
    archiveClassifier = 'sources'

    from sourceSets.main.allSource, sourceSets.test.allSource, sourceSets.main.resources.srcDirs
    exclude('**/*Test.java')
  }
```
So the sources JAR is composed by `main.allSource`, `test.allSource`, and `main.resources`. (BTW, we could just say `sourceSets.main.resources` instead of `sourceSets.main.resources.srcDirs`.)

The thing is, `allSource` [already includes `resources`](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.SourceSet.html#org.gradle.api.tasks.SourceSet:allSource). The default is a list `[java, resources]` as documented. So, it's redundant to specify `main.resources` again. Indeed, if I change it to
```
from sourceSets.main.allSource, sourceSets.test.allSource
```
I still see `dependencies.properties` in `build/libs/gax-2.7.1-SNAPSHOT-sources.jar` after running `./gradlew :gax:sourcesJar`. OTOH, if I change it to `from sourceSets.main.allJava, sourceSets.test.allSource`, I don't see the properties file in the `-sources.jar`.

Reply:
Thanks for the investigation. This clears things up. As a side note, I'm not sure why we are including test sources at all, yet exclude all files ending in `*Test.java`.

Reply:
Hmm... good point. My guess is that, this repo is also generating `*-testlib.jar` files that I think contain some test utility classes (whose filename doesn't end with `*Test.java`). Maybe per the Maven Central requirements or convention, because we are publishing `-testlib.jar` files, we also need to include source files.